File: 637-average-of-levels-in-binary-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class BTreeNode {
  val: number;
  left: BTreeNode | null;
  right: BTreeNode | null;
  constructor(val?: number, left?: BTreeNode | null, right?: BTreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function averageOfLevels(root: BTreeNode | null): number[] {
  const sumOfLevels: number[] = [];
  const cntOfLevels: number[] = [];

  const traversal = [{ node: root, level: 0 }];
  let idx = 0;
  while (!!traversal[idx]) {
    const { node, level } = traversal[idx];
    if (!sumOfLevels[level]) sumOfLevels[level] = 0;
    if (!cntOfLevels[level]) cntOfLevels[level] = 0;
    sumOfLevels[level] += node?.val ?? 0;
    cntOfLevels[level] += 1;

    if (!!node?.left) traversal.push({ node: node?.left, level: level + 1 });
    if (!!node?.right) traversal.push({ node: node?.right, level: level + 1 });
    idx++;
  }

  return sumOfLevels.map((s, i) => s / cntOfLevels[i]);
}

File: 637-average-of-levels-in-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traversal(vector<pair<double, int>>& sums, int depth, TreeNode* node) {
        if(node == NULL) return;
        if(depth >= sums.size()) sums.push_back({0, 0});
        sums[depth].first += (double)(node->val);
        sums[depth].second++;
        traversal(sums, depth + 1, node->left);
        traversal(sums, depth + 1, node->right);
    }

    vector<double> averageOfLevels(TreeNode* root) {
        vector<pair<double, int>> sums;
        traversal(sums, 0, root);
        vector<double> ans;
        for(auto x : sums) {
            ans.push_back((double)x.first / (double)x.second);
        }
        return ans;
    }
};
